### PR TITLE
Restructure classes

### PR DESCRIPTION
**Pull request by dennisylyung**
Adapt `Plan`, `Camp`, `Volunteer`, `Refugee` into persistent classes, and restructure their relationships. 
New structure:
ERD drawio

### CODE REVIEW

File: models/refugee.py [hunk 1]
```diff
@@ -92,30 +100,34 @@ def __sanitise_starting_date(self, starting_date: date):
...
                f"Number of Family Member: {self.num_of_family_member}\n" \
                f"Camp: {self.camp}\n" \
                f"Creation Date: {self.starting_date}\n" \
-               f"Medical Condition: {self.medical_condition_type}\n"\
-
+               f"Medical Condition: {self.medical_condition_type}\n"
 
     class InvalidNumOfFamilyMemberException(Exception):
         """
         Raise exception when number of family member entered.
```
Comment (Shalaka-Yerawadekar) [inline models/refugee.py:117]: Could this be corrected to: "Raise exception when the number of family member(s) entered is invalid."

File: models/tests/test_refugee.py [hunk 2]
```diff
@@ -40,64 +38,54 @@ def test_invalid_numOfFamilyMember1(self):
...
             Refugee(firstname="Tom",
                     lastname="Bond",
-                    camp=Camp("UCL"),
                     num_of_family_member=-10,
-                    starting_date= date(2020,1,2),
-                    medical_condition_type=Refugee.MedicalCondition.HIV)
+                    starting_date=date(2020, 1, 2),
+                    medical_condition_type=[Refugee.MedicalCondition.HIV])
 
     def test_invalid_firstname(self):
```
Comment (Shalaka-Yerawadekar) [inline models/tests/test_refugee.py:45]: Can we add: 

"""
Test invalid first name. 
"""

for consistency please? 

And similarly for the tests below.

File: models/base/document.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        def __init__(self, document):
+            super().__init__(f"Primary key not defined for {document.__class__.__name__}")
+
+    class PrimaryKeyNotSetError(Exception):
+        def __init__(self, field_name):
+            super().__init__(f"Primary key {field_name} not set")
+
+    class ReferrerNotFound(Exception):
+        def __init__(self, referrer_type=None, attribute_name=None):
+            super().__init__(f"Instanced is not referenced by "
```
Comment (Elynor-Kamil) [inline models/base/document.py:76]: Missing bracket and typo. I think this is meant to read:
```
super().__init__(f"Instance is not referenced by ")
```

Comment (Elynor-Kamil) [reply models/base/document.py:76]: I also realise that it's missing reference to the f string it's meant to pick up. I'm not sure whether this is meant to be Guessing it should be {referee} but maybe {field_name}?

Comment (Elynor-Kamil) [reply models/base/document.py:76]: No, wait, never mind - I missed the line after. All is well - resolving.

File: models/base/document.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+        self.save()
+
+    def __str__(self):
+        return f'{self.__class__.__name__}({self._data})'
+
+
+class IndexedDocument(Document, metaclass=MetaIndexedDocument):
+    """
+    Base class for all root level documents, directly persisted to disk.
+    A primary key must be defined to index all active documents.
```
Comment (Elynor-Kamil) [inline models/base/document.py:158]: Example for how to define this?

File: models/base/meta_document.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+
+class MetaDocument(type):
+    """
+    A metaclass for documents to register all fields and identify the primary key.
+    """
+
+    class MultiplePrimaryKeyError(Exception):
+        def __init__(self):
+            super().__init__(f"Only one primary key is allowed")
```
Comment (Elynor-Kamil) [inline models/base/meta_document.py:13]: What is 'f' referencing/returning here?

File: models/base/document.py (hunk 4 above)
Comment (dennisylyung) [reply models/base/document.py:158]: I think you meant to comment on #27. 
I have added an example to the docstring.

File: models/base/meta_document.py (hunk 5 above)
Comment (dennisylyung) [reply models/base/meta_document.py:13]: Good catch, it's useless. 
(I think you meant to comment on #27 though, i will update there)

**Review by YunsyYin — APPROVED** [review]
The functions in the Volunteer class were already tested successfully

**Review by YingboZhou — APPROVED** [review]

**Review by Elynor-Kamil — APPROVED** [review]

**Review by Shalaka-Yerawadekar — APPROVED** [review]

**Review by michellechan1123 — COMMENTED** [review]
just wondering why do we need to reload when closing camp

```
def close_camps(self, *camps: Camp) -> None:
        """
        Close one or more camps.
        MissingCampsError is raised if the call will close all camps in the plan.
        CampNotFoundError is raised if one or more camps are not found in the plan.
        """
        for camp in camps:
            if camp not in self.camps:
                raise self.CampNotFoundError(camp)
            self.camps.remove(camp)
        if len(self.camps) == 0:
            self.reload()
            raise self.MissingCampsError()
        else:
            self.save()
```

**Review by YingboZhou — APPROVED** [review]

**Comment by dennisylyung** [comment]
> just wondering why do we need to reload when closing camp
> 
> ```
> def close_camps(self, *camps: Camp) -> None:
>         """
>         Close one or more camps.
>         MissingCampsError is raised if the call will close all camps in the plan.
>         CampNotFoundError is raised if one or more camps are not found in the plan.
>         """
>         for camp in camps:
>             if camp not in self.camps:
>                 raise self.CampNotFoundError(camp)
>             self.camps.remove(camp)
>         if len(self.camps) == 0:
>             self.reload()
>             raise self.MissingCampsError()
>         else:
>             self.save()
> ```
 
It is not permitted to close all camps in a plan. 
If a user attempts to close the last camp, we will rollback the whole `close_camps` call by restoring the last saved plan state from disk.

**Pull request merged**